### ISSUE TITLE
Containers cannot access internet (outbound)

### ISSUE THREAD

**Issue by Nepoxx** [moby/moby#36151]
I cannot access the internet from my containers, however inbound connections still work. As far as I know the problem occurred without human intervention ("it was working fine and now it isn't")

**Steps to reproduce the issue:**
```
nepoxx@host:~$ ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=59 time=12.7 ms
...
^C
--- 8.8.8.8 ping statistics ---
6 packets transmitted, 6 received, 0% packet loss, time 5007ms
rtt min/avg/max/mdev = 12.683/12.834/13.345/0.239 ms


nepoxx@host:~$ docker run -i --rm alpine ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8): 56 data bytes
^C
--- 8.8.8.8 ping statistics ---
8 packets transmitted, 0 packets received, 100% packet loss
```

I had this issue with docker 1.13, I've upgraded to 17.12.0-ce in the hopes that it would fix it but it didn't help. I've restart docker, the host, I've tried:
```
pkill docker
iptables -t nat -F
ifconfig docker0 down
brctl delbr docker0
docker -d
```
to no avail

**Output of `docker version`:**

```
Docker version 17.12.0-ce, build c97c6d6
```

**Output of `docker info`:**

```
Containers: 3
 Running: 2
 Paused: 0
 Stopped: 1
Images: 13
Server Version: 17.12.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 74
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 89623f28b87a6004d4b785663257362d1658a729
runc version: b2567b37d7b75eb4cf325b77297b140ea686ce8f
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-112-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.85GiB
Name: vqlnxlivechat
ID: K4EK:CQHO:MGWH:HFFS:IWHS:NIAN:WMXE:6VHX:6AM3:CPQ5:A3AD:6GS3
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Linux host 4.4.0-112-generic #135-Ubuntu SMP Fri Jan 19 11:48:36 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

**Comment by duraki** [comment]
Same issue here, so before opening new one, have you managed to solve this?

**Comment by expz** [comment]
After three years of using docker without problems, this issue is affecting me today. The only two new things at the time it showed up are (1) used a debian container for the first time, (2) added my user to the docker group and began using docker without sudo.

Attempted fixes which did not work:

0. `sudo systemctl restart docker`

1. Toggle `--iptables` dockerd switch between true and false, restarting docker service between toggles.

2. Adding VIF using `sudo brctl addif docker0 <vif>`.

3. Forcing rebuild of bridge
```
sudo systemctl stop docker
sudo pkill docker
sudo iptables -t nat -F
sudo ifconfig docker0 down
sudo brctl delbr docker0
sudo systemctl start docker
```

I am using docker `version 17.12.1-ce, build 7390fc6` on Ubuntu 17.10.

**Comment by expz** [comment]
For anyone running into this, rebooting the OS fixed it.

**Comment by through-a-haze** [comment]
Have anyone solved the problem? Rebooting the OS didn't help and I can't use --net=host because of linked containers.

**Comment by duraki** [comment]
Check your gateways with `arp` and see if any matches your Docker IP. If it does, remove the gateway with `route del` and try restarting docker engine. I can't remember how I fixed this issue.

**Comment by expz** [comment]
In my case, the reboot of the OS probably solved the problem because it killed the VPN connection and restarted the docker service. There is some conflict between my VPN connection and the docker networking setup. You could also try the four methods I listed.

**Comment by gboyegadada** [comment]
Lost my whole day to this. A reboot (Win 10) worked in the end. Thanks @expz

**Comment by trinhpham** [comment]
Here is what I experienced on Windows Server 2016.
Docker in with Linux mode.
If i understand correctly, the Moby linux recognizes and uses **wrong network interface** (**eth0** instead of **hvint0**), so neither the MobyVM nor its containers can access to the Internet.
Please find the detail information below.

# The environment:
![image](https://user-images.githubusercontent.com/9128061/42086225-a89e1d7e-7bbc-11e8-9b91-cc13d760744d.png)
![image](https://user-images.githubusercontent.com/9128061/42086300-e63b35c2-7bbc-11e8-9296-ef3e8d19537e.png)

# docker version
```Powershell
PS C:\Users\trinh.pham> docker version
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Thu Apr 26 07:12:48 2018
 OS/Arch:      windows/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Thu Apr 26 07:22:38 2018
  OS/Arch:      linux/amd64
  Experimental: true
```

# Docker info output
```Powershell
PS C:\Users\trinh.pham> docker info
Containers: 11
 Running: 11
 Paused: 0
 Stopped: 0
Images: 14
Server Version: 18.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.87-linuxkit-aufs
Operating System: Docker for Windows
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 9.725GiB
Name: linuxkit-00155da7cb03
ID: NS7F:BLDY:L2ZU:UJ7T:N4RF:MN6R:BPBR:WGEY:RFBL:DBE3:T2GD:EQ5H
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 107
 Goroutines: 102
 System Time: 2018-06-29T09:53:24.277022Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.0/8
Registry Mirrors:
Live Restore Enabled: false
```

# docker run --rm --net=host alpine ip addr
```Powershell
PS C:\Users\trinh.pham> docker run --rm --net=host alpine ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 brd 127.255.255.255 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: hvint0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP qlen 1000
    link/ether 00:15:5d:a7:cb:03 brd ff:ff:ff:ff:ff:ff
    inet 10.0.75.2/24 scope global hvint0
       valid_lft forever preferred_lft forever
    inet6 fe80::215:5dff:fea7:cb03/64 scope link
       valid_lft forever preferred_lft forever
3: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN qlen 1
    link/ipip 0.0.0.0 brd 0.0.0.0
4: ip6tnl0@NONE: <NOARP> mtu 1452 qdisc noop state DOWN qlen 1
    link/tunnel6 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
5: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UNKNOWN qlen 1000
    link/ether 02:50:00:00:00:01 brd ff:ff:ff:ff:ff:ff
    inet 192.168.65.3/24 brd 192.168.65.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::f6cf:e3b8:f0f9:408f/64 scope link
       valid_lft forever preferred_lft forever
6: br-2b5472538d72: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP
    link/ether 02:42:d2:7b:9c:57 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.1/16 brd 172.18.255.255 scope global br-2b5472538d72
       valid_lft forever preferred_lft forever
    inet6 fe80::42:d2ff:fe7b:9c57/64 scope link
       valid_lft forever preferred_lft forever
7: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN
    link/ether 02:42:d3:bd:d2:56 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:d3ff:febd:d256/64 scope link
       valid_lft forever preferred_lft forever
```

# docker run --rm --net=host alpine route
```Powershell
PS C:\Users\trinh.pham> docker run --rm --net=host alpine route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         192.168.65.1    0.0.0.0         UG    205    0        0 eth0
10.0.75.0       *               255.255.255.0   U     0      0        0 hvint0
127.0.0.0       *               255.0.0.0       U     0      0        0 lo
172.17.0.0      *               255.255.0.0     U     0      0        0 docker0
172.18.0.0      *               255.255.0.0     U     0      0        0 br-2b5472538d72
192.168.65.0    *               255.255.255.0   U     205    0        0 eth0
```

# docker run --rm alpine ip addr
```Powershell
PS C:\Users\trinh.pham> docker run --rm alpine ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN qlen 1
    link/ipip 0.0.0.0 brd 0.0.0.0
3: ip6tnl0@NONE: <NOARP> mtu 1452 qdisc noop state DOWN qlen 1
    link/tunnel6 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
120: eth0@if121: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.2/16 brd 172.17.255.255 scope global eth0
       valid_lft forever preferred_lft forever
```

# docker run --rm alpine route
```Powershell
PS C:\Users\trinh.pham> docker run --rm alpine route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         172.17.0.1      0.0.0.0         UG    0      0        0 eth0
172.17.0.0      *               255.255.0.0     U     0      0        0 eth0
```

**Comment by DMeechan** [comment]
Has anyone managed to solve this any other way?

I've tried restarting the Docker service, rebooting the OS, and forcing a bridge rebuild. No success so far.

I'm running `Ubuntu 18.10 (Pop!_OS 18.10)` and `Docker version 18.06.1-ce, build e68fc7a`.

**Comment by prakashs-treselle** [comment]
I too have same issue with AWS workspace (AWS Linux 2), 
Docker works fine with --host
I tried to update /etc/resolve
added /etc/docker/daemon.json

Still not working, If any one resolved the issue, Please do notify

**Comment by jobinau** [comment]
I am also facing same problem on Ubuntu 18.04.
No Internent without --net=host option.

**Comment by cur3n4** [comment]
Enabling ip forwarding (which had become suddenly disabled in the host) worked for me. 
`sudo sysctl net.ipv4.ip_forward=1`

**Comment by olljanat** [comment]
I'm sorry to say this but it really sounds that you are seeing this issue because you are using non-supported configuration(s).

@trinhpham combination of Windows Server and CE version of Docker is not supported. Please use either Win 10 with Docker CE (if you need Linux support) or [Win Srv with enterprise edition](https://docs.microsoft.com/en-us/virtualization/windowscontainers/quick-start/quick-start-windows-server).


@DMeechan I have ever heard about Pop!_OS but I guess that it causes your problem. Please, test with normal version of Ubuntu and [latest Docker-CE version](https://docs.docker.com/install/linux/docker-ce/ubuntu/)


@prakashs-treselle I think that AWS Linux 2 is not supported Docker host OS. How ever you can run it [inside of Docker](https://hub.docker.com/_/amazonlinux).


@jobinau please check that you don't have any special iptables rules. Docker uses iptables to provide networking for containers.


Here is example from Ubuntu box with standard config and latest version of Docker:
```bash
$ docker run --rm alpine ping -c 2 www.google.com
PING www.google.com (216.58.207.228): 56 data bytes
64 bytes from 216.58.207.228: seq=0 ttl=54 time=21.100 ms
64 bytes from 216.58.207.228: seq=1 ttl=54 time=20.647 ms

--- www.google.com ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 20.647/20.873/21.100 ms
```

And here is same from Windows Server 2019 with latest Docker version on Azure:
```powershell
PS C:\> docker run --rm mcr.microsoft.com/windows/nanoserver:1809 ping -n 2 www.google.com

Pinging www.google.com [216.58.211.100] with 32 bytes of data:
Reply from 216.58.211.100: bytes=32 time=1ms TTL=50
Reply from 216.58.211.100: bytes=32 time=1ms TTL=50

Ping statistics for 216.58.211.100:
    Packets: Sent = 2, Received = 2, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 1ms, Maximum = 1ms, Average = 1ms
```

**Comment by Ethan3600** [comment]
I'm also seeing this issue on Ubuntu 16 & 18 LTS:

```bash
 ▲ ~ docker run --rm alpine ping -c 2 www.google.com
Unable to find image 'alpine:latest' locally
latest: Pulling from library/alpine
9d48c3bd43c5: Pull complete
Digest: sha256:72c42ed48c3a2db31b7dafe17d275b634664a708d901ec9fd57b1529280f01fb
Status: Downloaded newer image for alpine:latest
ping: bad address 'www.google.com'
```

Same things with different image:

```bash
 ▲ ~ docker run --network bridge --rm -it ubuntu:trusty bash                                                                                     master 3d :: ⬡
root@d846d33bcbd3:/# ping google.com
ping: unknown host google.com
root@d846d33bcbd3:/#

```

Docker Version:

```bash
 △ ~ docker version                                                                                                                                                                       12s
Client:
 Version:           18.09.7
 API version:       1.39
 Go version:        go1.10.1
 Git commit:        2d0083d
 Built:             Fri Aug 16 14:20:06 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.7
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.1
  Git commit:       2d0083d
  Built:            Wed Aug 14 19:41:23 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

Things I've attempted to fix this issue with no luck:

1. I've removed docker0 and restarted docker (daemon)
2. [Ensured that the allocated ip addresses for docker don't conflict with any DNS IPs](https://superuser.com/questions/1336567/installing-docker-ce-in-ubuntu-18-04-breaks-internet-connectivity-of-host)
3. Upgraded from Ubuntu 16.04 to Ubuntu 18.04
4. Toggled the `--iptables` flag 
5. Made the DNS servers 8.8.8.8 and 8.8.4.4 explicitly in /etc/docker/daemon.json
6. flushed iptables `iptables -F`
7. Ensured sudo sysctl net.ipv4.ip_forward is set to 1

Note, this is happening out of the box. The first thing I did after I setup my new Ubuntu 18 OS was install docker / docker-compose and ran `docker run --rm alpine ping -c 2 www.google.com`.

Also wanted to point out that using the `host` network driver allows me to make any requests to the outside world. But, using the default bridge or a custom network (with the bridge driver) results in no access to the internet.

Really hoping this can be resolved :pray:

**Comment by dimovnike** [comment]
Same issue is seen on centos7 with 19.03.3. It is not DNS because pings to 1.1.1.1/8.8.8.8 do not work either.

**Comment by bleed2002** [comment]
Having the same issue on one frech installed CentOS 7.7 minimal (Hetzner Server).
Tested with docker 18.09.9 and latest 19.03.3. No custom configuration for CentOS and docker.

I tried basically everything, but no luck. Very very strange, since I manage dozens of servers with the same stack and configuration (ansible provisioned), not having this problem??

On this server (Type EX 42 - https://www.hetzner.de/dedicated-rootserver/ex42) the problem is reproducible:
- Minimal CentOS 7.6 install (by Hetzner installimage)
- yum update (also tried without it)
- docker installation:
`yum install -y yum-utils
yum-config-manager --add-repo
https://download.docker.com/linux/centos/docker-ce.repo
yum install docker-ce docker-ce-cli containerd.io
systemctl start docker`
- Tests:
`# does not work
docker run -it --rm alpine ping 1.1.1.1

# works
docker run -it --rm --net=host alpine ping 1.1.1.1`

`
# uname -a
3.10.0-1062.1.2.el7.x86_64 #1 SMP Mon Sep 30 14:19:46 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

# cat /etc/redhat-release 
CentOS Linux release 7.7.1908 (Core)

# docker version
Client: Docker Engine - Community
 Version:           19.03.3
 API version:       1.40
 Go version:        go1.12.10
 Git commit:        a872fc2f86
 Built:             Tue Oct  8 00:58:10 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.3
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.10
  Git commit:       a872fc2f86
  Built:            Tue Oct  8 00:56:46 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683`

Any ideas? Thanks!

**Comment by bleed2002** [comment]
Solved it for me:

Since systemd version >=219 ip forwarding (net.ipv4.conf.<interface>.forwarding) is off by default.
To activate I had to place the option IPForward=true in the respective systemd network config file:

```
[Network]
...
IPForward=true
```

After restart of systemd-networkd (`systemctl restart systemd-networkd`) everything works as expected. 

Reference: https://docs.docker.com/v17.09/engine/installation/linux/linux-postinstall/#ip-forwarding-problems

**Comment by D1plomat** [comment]
> Solved it for me:
> 
> Since systemd version >=219 ip forwarding (net.ipv4.conf.**interfaceName**.forwarding) is off by default.
> To activate I had to place the option IPForward=true in the respective systemd network config file:
> 
> ```
> [Network]
> ...
> IPForward=true
> ```
> 
> After restart of systemd-networkd (`systemctl restart systemd-networkd`) everything works as expected.
> 
> Reference: https://docs.docker.com/v17.09/engine/installation/linux/linux-postinstall/#ip-forwarding-problems

I tried this, and all of the replies above and none of it worked. I'm running manjaro linux. I don't use dnsmasq, ping from host and from container to 8.8.8.8 works perfectly but when running nslookup I get thath there is no route to host from my default gateway which is router that ISP provided

example:
````
>>>>> docker run busybox nslookup google.com                                                                                                                                                                                                                                                      
;; connection timed out; no servers could be reached

nslookup: write to '192.168.0.1': No route to host

````
docker info output:
````
Client:
 Debug Mode: false

Server:
 Containers: 75
  Running: 0
  Paused: 0
  Stopped: 75
 Images: 14
 Server Version: 19.03.3-ce
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: d50db0a42053864a270f648048f9a8b4f24eced3.m
 runc version: d736ef14f0288d6993a1845745d6756cfc9ddd5a
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.2.20-1-MANJARO
 Operating System: Manjaro Linux
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 31.34GiB
 Name: filip-pc
 ID: DJFI:HJJJ:PZQX:SNP5:KXVB:FYCO:NOGK:WJGI:6GNP:IP47:2MFM:H2NR
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
````

ifconfig output:

````
br-e4c22441e9b9: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 172.19.0.1  netmask 255.255.0.0  broadcast 172.19.255.255
        ether 02:42:ff:3e:c6:4e  txqueuelen 0  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

docker0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 192.168.1.5  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 fe80::42:b6ff:fe00:1e1d  prefixlen 64  scopeid 0x20<link>
        ether 02:42:b6:00:1e:1d  txqueuelen 0  (Ethernet)
        RX packets 4  bytes 196 (196.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 7  bytes 666 (666.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

enp0s31f6: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.0.135  netmask 255.255.255.0  broadcast 192.168.0.255
        inet6 fe80::dd73:de5c:9d11:8462  prefixlen 64  scopeid 0x20<link>
        ether 2c:fd:a1:e2:04:0c  txqueuelen 1000  (Ethernet)
        RX packets 15146  bytes 14332560 (13.6 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 9062  bytes 1452665 (1.3 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device interrupt 16  memory 0xdf600000-df620000  

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
``
``

**Comment by yogeshkathayat** [comment]
for me Rebooting the OS worked.

**Comment by iNevd** [comment]
> Having the same issue on one frech installed CentOS 7.7 minimal (Hetzner Server).
> Tested with docker 18.09.9 and latest 19.03.3. No custom configuration for CentOS and docker.
> 
> I tried basically everything, but no luck. Very very strange, since I manage dozens of servers with the same stack and configuration (ansible provisioned), not having this problem??
> 
> On this server (Type EX 42 - https://www.hetzner.de/dedicated-rootserver/ex42) the problem is reproducible:
> 
> 1. Minimal CentOS 7.6 install (by Hetzner installimage)
> 2. yum update (also tried without it)
> 3. docker installation:
>    `yum install -y yum-utils yum-config-manager --add-repo https://download.docker.com/linux/centos/docker-ce.repo yum install docker-ce docker-ce-cli containerd.io systemctl start docker`
> 4. Tests:
> 
> `docker run -it --rm alpine ping 1.1.1.1`
> _> does NOT work
> 
> `docker run -it --rm --net=host alpine ping 1.1.1.1`
> _> DOES work
> 
> ## Additional info
> uname -a
> 3.10.0-1062.1.2.el7.x86_64 #1 SMP Mon Sep 30 14:19:46 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
> 
> cat /etc/redhat-release
> CentOS Linux release 7.7.1908 (Core)
> 
> docker version
> Client: Docker Engine - Community
> Version: 19.03.3
> API version: 1.40
> Go version: go1.12.10
> Git commit: a872fc2f86
> Built: Tue Oct 8 00:58:10 2019
> OS/Arch: linux/amd64
> Experimental: false
> 
> Server: Docker Engine - Community
> Engine:
> Version: 19.03.3
> API version: 1.40 (minimum version 1.12)
> Go version: go1.12.10
> Git commit: a872fc2f86
> Built: Tue Oct 8 00:56:46 2019
> OS/Arch: linux/amd64
> Experimental: false
> containerd:
> Version: 1.2.10
> GitCommit: b34a5c8af56e510852c35414db4c1f4fa6172339
> runc:
> Version: 1.0.0-rc8+dev
> GitCommit: 3e425f80a8c931f88e6d94a8c831b9d5aa481657
> docker-init:
> Version: 0.18.0
> GitCommit: fec3683
> 
> Any ideas? Thanks!

try `systemctl stop firewalld` is ok for me

**Comment by BenjaminBini** [comment]
Exact same issue, tried everything, still no internet from the container.
CentOS 7.
```
Client:
 Debug Mode: false

Server:
 Containers: 27
  Running: 0
  Paused: 0
  Stopped: 27
 Images: 9
 Server Version: 19.03.6
 Storage Driver: devicemapper
  Pool Name: docker-253:0-135106576-pool
  Pool Blocksize: 65.54kB
  Base Device Size: 10.74GB
  Backing Filesystem: xfs
  Udev Sync Supported: true
  Data file: /dev/loop0
  Metadata file: /dev/loop1
  Data loop file: /var/lib/docker/devicemapper/devicemapper/data
  Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
  Data Space Used: 1.371GB
  Data Space Total: 107.4GB
  Data Space Available: 37.14GB
  Metadata Space Used: 2.908MB
  Metadata Space Total: 2.147GB
  Metadata Space Available: 2.145GB
  Thin Pool Minimum Free Space: 10.74GB
  Deferred Removal Enabled: true
  Deferred Deletion Enabled: true
  Deferred Deleted Device Count: 0
  Library Version: 1.02.158-RHEL7 (2019-05-13)
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: b34a5c8af56e510852c35414db4c1f4fa6172339
 runc version: 3e425f80a8c931f88e6d94a8c831b9d5aa481657
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 3.10.0-327.el7.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 3.703GiB
 Name: sv124st.sully-group.fr
 ID: PFCS:N6PF:2M5Z:VFSI:IK66:RRYB:GP5U:UU5R:ICX2:YSHP:2ZMR:DULU
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: the devicemapper storage-driver is deprecated, and will be removed in a future release.
WARNING: devicemapper: usage of loopback devices is strongly discouraged for production use.
         Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.

```

```
Linux xxx.aaa-bbb.fr 3.10.0-327.el7.x86_64 #1 SMP Thu Nov 19 22:10:57 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by Kinark** [comment]
Same issue here. Tried everything related above :(

**Comment by rotulet** [comment]
@Kinark the same! I have update my docker and whole system. 

```
$ docker version
Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b7f0
 Built:             Wed Mar 11 01:25:56 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       afacb8b7f0
  Built:            Wed Mar 11 01:24:28 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```
```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux bullseye/sid
Release:	8.11
Codename:	jessie
```

May be an AppArmor retriction or specific network configuration ipv4/ipv6?

**Comment by BenjaminBini** [comment]
Solved for me: SELinux was enabled on the virtual machine.

Disabling it solved the problem: https://linuxize.com/post/how-to-disable-selinux-on-centos-7/

**Comment by tinti** [comment]
Testing with `--privileged` worked for me. Any ideas?

<pre>
Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b
 Built:             Wed Mar 11 01:27:04 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       afacb8b
  Built:            Wed Mar 11 01:25:42 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
</pre>

<pre>
CentOS Linux release 7.7.1908 (Core)
</pre>

**Comment by tinti** [comment]
> Testing with `--privileged` worked for me. Any ideas?
> 
> Client: Docker Engine - Community
>  Version:           19.03.8
>  API version:       1.40
>  Go version:        go1.12.17
>  Git commit:        afacb8b
>  Built:             Wed Mar 11 01:27:04 2020
>  OS/Arch:           linux/amd64
>  Experimental:      false
> 
> Server: Docker Engine - Community
>  Engine:
>   Version:          19.03.8
>   API version:      1.40 (minimum version 1.12)
>   Go version:       go1.12.17
>   Git commit:       afacb8b
>   Built:            Wed Mar 11 01:25:42 2020
>   OS/Arch:          linux/amd64
>   Experimental:     false
>  containerd:
>   Version:          1.2.13
>   GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
>  runc:
>   Version:          1.0.0-rc10
>   GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
>  docker-init:
>   Version:          0.18.0
>   GitCommit:        fec3683
> 
> CentOS Linux release 7.7.1908 (Core)

Turns out that only `--cap-add SYS_PTRACE` is enough in my case.

I am running **Wine** inside **Docker**.

**Comment by Kinark** [comment]
After A LOT of time trying to solve this issue I think I finally found a solution in an ebook I found on the internet.

Here's the steps I took to solve the issue:
```
$ sudo service docker stop
$ sudo nano /etc/default/docker
```
Edit the file and uncomment the DOCKER_OPTS option modifying it to somthing like this:
`DOCKER_OPTS="--dns 8.8.8.8 --dns 8.8.4.4 --iptables=false --ip-forward=false"`
```
$ sudo su
$ echo 1 > /proc/sys/net/ipv4/ip_forward
$ iptables -t nat -A POSTROUTING -s 172.17.0.0/16 -j MASQUERADE
$ service docker restart
$ exit
```
Now hopefully your containers have access to the internet again.

I hope to help someone :)

**Comment by otidh** [comment]
> After A LOT of time trying to solve this issue I think I finally found a solution in an ebook I found on the internet.
> 
> Here's the steps I took to solve the issue:
> 
> ```
> $ sudo service docker stop
> $ sudo nano /etc/default/docker
> ```
> 
> Edit the file and uncomment the DOCKER_OPTS option modifying it to somthing like this:
> `DOCKER_OPTS="--dns 8.8.8.8 --dns 8.8.4.4 --iptables=false --ip-forward=false"`
> 
> ```
> $ sudo su
> $ echo 1 > /proc/sys/net/ipv4/ip_forward
> // Add the IP range you want to have access to the internet
> $ iptables -t nat -A POSTROUTING -s 172.17.0.0/16 -j MASQUERADE
> $ service docker restart
> $ exit
> ```
> 
> Now hopefully your containers have access to the internet again.
> 
> I hope to help someone :)

Thank you!! Your solution works for me!

**Comment by RooSoft** [comment]
I've had a similar networking problem and found a solution...

Under an hypervisor, I had made a template out of a plain Docker VM. I then cloned it into 3 different hosts, clearly sharing the exact same config and data, at least at first boot.

From there on, I started having containers unable to reach the internet.

The solution? Wipe all configs and data.

```bash
systemctl stop docker
rm -fr /var/lib/docker
rm -fr /etc/docker
systemctl start docker
```

Seems like Docker re-configures itself in that case. The next thing I've realized, my containers started working flawlessly.

**Comment by KamalAman** [comment]
For those warriors out there, I had a similar issue. 

The resolution for me was the network I was using did not allow UDP traffic to the DNS which is sometimes required by DNS servers to perform lookups. 

`nslookup google.com` on the host machine on Ubuntu uses a caching server 127.0.0.57 which makes TCP requests to the DNS worked. So everything seemed okay, but an `nslookup` within the container with `--dns 8.8.8.8` run time argument did not work. It turns out that the default lookup is actually UDP and not TCP. So when i ran `nslookup google.com 8.8.8.8` on the host machine, it failed. The network needed to be fixed to resolve this issue. 

I hope this may help

**Comment by rkok** [comment]
Suddenly started happening for me as well, on Ubuntu 20.04.

What fixed it in my case:
```
systemctl restart networking
systemctl restart NetworkManager
systemctl restart docker
```

`net.ipv4.ip_forward` was already enabled.

**Comment by skududufru** [comment]
> sudo sysctl net.ipv4.ip_forward=1

What the hell!
It was that thank you!

**Comment by colomonkey** [comment]
I tried most of the solutions above with no luck. This is happening on a docker host running Ubuntu 20.04 with swarm enabled. I am using --network host works but no other solution exists for accessing the internet via IP (Not talking about DNS). Connectivity on the host itself works fine, but not via a container with a bridged or overlay network.

Tried:
- rebooting host or restarting docker daemon
- toggle iptables=true / false
- set forwarding net.ipv4.ip_forward=1 permanent <- it is.
- run with --privileged 
- no network manager of firewalld running

General info: 

`uname -a: Linux svc-1.speaksee.ai 5.4.0-1028-aws #29-Ubuntu SMP Mon Oct 5 15:30:10 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`

docker version
```
Client: Docker Engine - Community
 Version:           19.03.12
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        48a66213fe
 Built:             Mon Jun 22 15:45:36 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.12
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       48a66213fe
  Built:            Mon Jun 22 15:44:07 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

iptables on host:

```
# Generated by iptables-save v1.8.4 on Tue Oct 27 13:08:09 2020
*mangle
:PREROUTING ACCEPT [21277:4051299]
:INPUT ACCEPT [20297:3860588]
:FORWARD ACCEPT [980:190711]
:OUTPUT ACCEPT [18874:3960389]
:POSTROUTING ACCEPT [19854:4151100]
COMMIT
# Completed on Tue Oct 27 13:08:09 2020
# Generated by iptables-save v1.8.4 on Tue Oct 27 13:08:09 2020
*filter
:INPUT ACCEPT [11851:1923824]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [10891:1921544]
:DOCKER - [0:0]
:DOCKER-INGRESS - [0:0]
:DOCKER-ISOLATION-STAGE-1 - [0:0]
:DOCKER-ISOLATION-STAGE-2 - [0:0]
:DOCKER-USER - [0:0]
-A FORWARD -j DOCKER-USER
-A FORWARD -j DOCKER-INGRESS
-A FORWARD -j DOCKER-ISOLATION-STAGE-1
-A FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -o docker0 -j DOCKER
-A FORWARD -i docker0 ! -o docker0 -j ACCEPT
-A FORWARD -i docker0 -o docker0 -j ACCEPT
-A FORWARD -o br-b123e2e118d9 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -o br-b123e2e118d9 -j DOCKER
-A FORWARD -i br-b123e2e118d9 ! -o br-b123e2e118d9 -j ACCEPT
-A FORWARD -i br-b123e2e118d9 -o br-b123e2e118d9 -j ACCEPT
-A FORWARD -o docker_gwbridge -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -o docker_gwbridge -j DOCKER
-A FORWARD -i docker_gwbridge ! -o docker_gwbridge -j ACCEPT
-A FORWARD -i docker_gwbridge -o docker_gwbridge -j DROP
-A DOCKER-INGRESS -p tcp -m tcp --dport 443 -j ACCEPT
-A DOCKER-INGRESS -p tcp -m state --state RELATED,ESTABLISHED -m tcp --sport 443 -j ACCEPT
-A DOCKER-INGRESS -p tcp -m tcp --dport 80 -j ACCEPT
-A DOCKER-INGRESS -p tcp -m state --state RELATED,ESTABLISHED -m tcp --sport 80 -j ACCEPT
-A DOCKER-INGRESS -j RETURN
-A DOCKER-ISOLATION-STAGE-1 -i docker0 ! -o docker0 -j DOCKER-ISOLATION-STAGE-2
-A DOCKER-ISOLATION-STAGE-1 -i br-b123e2e118d9 ! -o br-b123e2e118d9 -j DOCKER-ISOLATION-STAGE-2
-A DOCKER-ISOLATION-STAGE-1 -i docker_gwbridge ! -o docker_gwbridge -j DOCKER-ISOLATION-STAGE-2
-A DOCKER-ISOLATION-STAGE-1 -j RETURN
-A DOCKER-ISOLATION-STAGE-2 -o docker0 -j DROP
-A DOCKER-ISOLATION-STAGE-2 -o br-b123e2e118d9 -j DROP
-A DOCKER-ISOLATION-STAGE-2 -o docker_gwbridge -j DROP
-A DOCKER-ISOLATION-STAGE-2 -j RETURN
-A DOCKER-USER -j RETURN
COMMIT
# Completed on Tue Oct 27 13:08:09 2020
# Generated by iptables-save v1.8.4 on Tue Oct 27 13:08:09 2020
*nat
:PREROUTING ACCEPT [697:67620]
:INPUT ACCEPT [697:67620]
:OUTPUT ACCEPT [730:70152]
:POSTROUTING ACCEPT [737:70512]
:DOCKER - [0:0]
:DOCKER-INGRESS - [0:0]
-A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER-INGRESS
-A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER
-A OUTPUT -m addrtype --dst-type LOCAL -j DOCKER-INGRESS
-A OUTPUT ! -d 127.0.0.0/8 -m addrtype --dst-type LOCAL -j DOCKER
-A POSTROUTING -s 172.17.0.0/16 ! -o docker0 -j MASQUERADE
-A POSTROUTING -o docker_gwbridge -m addrtype --src-type LOCAL -j MASQUERADE
-A POSTROUTING -s 192.168.88.0/24 ! -o br-b123e2e118d9 -j MASQUERADE
-A POSTROUTING -s 172.18.0.0/16 ! -o docker_gwbridge -j MASQUERADE
-A DOCKER -i docker0 -j RETURN
-A DOCKER -i br-b123e2e118d9 -j RETURN
-A DOCKER -i docker_gwbridge -j RETURN
-A DOCKER-INGRESS -p tcp -m tcp --dport 443 -j DNAT --to-destination 172.18.0.2:443
-A DOCKER-INGRESS -p tcp -m tcp --dport 80 -j DNAT --to-destination 172.18.0.2:80
-A DOCKER-INGRESS -j RETURN
COMMIT
```
ifconfig on host:

```
br-b123e2e118d9: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 192.168.88.1  netmask 255.255.255.0  broadcast 192.168.88.255

docker0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 172.17.255.255

docker_gwbridge: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.18.0.1  netmask 255.255.0.0  broadcast 172.18.255.255


ens5: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 9001 <- external
        inet 10.0.1.148  netmask 255.255.255.0  broadcast 10.0.1.255

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
```

route in image with user defined network:

```
docker run -it --network my-net --rm nginx:alpine route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         192.168.88.1    0.0.0.0         UG    0      0        0 eth0
192.168.88.0    *               255.255.255.0   U     0      0        0 eth0
```

Network in that image:

```
docker run -it --network my-net --rm nginx:alpine ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
289: eth0@if290: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether 02:42:c0:a8:58:02 brd ff:ff:ff:ff:ff:ff
    inet 192.168.88.2/24 brd 192.168.88.255 scope global eth0
       valid_lft forever preferred_lft forever
```
The ping tool reports 100% loss, meaning that the traffic is accepted somewhere but not returned correctly?

**Comment by michelescarlato** [comment]
It starts to happen to me while I was connecting via wifi (I usually work with eth) to a network 172.17.2.0/16.
Docker bridge configured (by default) on 172.17.0.0/16.
The workaround has been to modify the subnet mask of the docker bridge (I set it to /24).

Hope this can help.

**Comment by zaverden** [comment]
I don't know what kind of dark magic it was but solution in https://github.com/moby/moby/issues/26567#issuecomment-255517029 helped me.
> Toggle iptables=false to true; restarting dockerd
> Toggle iptables=true to false; restarting dockerd

I toggled iptables in `/etc/docker/daemon.json`

Host: Ubuntu 20.04.1 LTS
`docker version`:
```
Client: Docker Engine - Community
 Version:           19.03.13
 API version:       1.40
 Go version:        go1.13.15
 Git commit:        4484c46d9d
 Built:             Wed Sep 16 17:02:52 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.13
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       4484c46d9d
  Built:            Wed Sep 16 17:01:20 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.7
  GitCommit:        8fba4e9a7d01810a393d5d25a3621dc101981175
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Comment by zhanghe9702** [comment]
**i have tried all methods, not work for me**
scrapy@node1:~$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 20.10.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc version: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.8.0-34-generic
 Operating System: Ubuntu 20.10
 OSType: linux
 Architecture: x86_64
 CPUs: 24
 Total Memory: 62.78GiB
 Name: node1
 ID: VPVV:KYSU:KZV2:4I3J:SAJE:MGZN:262D:KIPH:MBBL:GCT3:A35G:SHUU
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: zhanghe9702
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio weight support
WARNING: No blkio weight_device support

**interface info**

scrapy@node1:~$ ifconfig -a
br-2afebfb6e7a2: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 172.18.0.1  netmask 255.255.0.0  broadcast 172.18.255.255
        inet6 fe80::42:f8ff:fef3:22ed  prefixlen 64  scopeid 0x20<link>
        ether 02:42:f8:f3:22:ed  txqueuelen 0  (以太网)
        RX packets 78  bytes 4052 (4.0 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 99  bytes 16643 (16.6 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

br-8897ffe2f381: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 192.168.49.1  netmask 255.255.255.0  broadcast 192.168.49.255
        ether 02:42:62:06:d2:4c  txqueuelen 0  (以太网)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

docker0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 172.17.255.255
        inet6 fe80::2c35:f0df:443f:6dd8  prefixlen 64  scopeid 0x20<link>
        ether 02:42:35:4b:3e:48  txqueuelen 0  (以太网)
        RX packets 2322  bytes 182431 (182.4 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 211  bytes 17077 (17.0 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

enp5s0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.50.76  netmask 255.255.255.0  broadcast 192.168.50.255
        inet6 fe80::98e2:4ad1:8cc7:4f26  prefixlen 64  scopeid 0x20<link>
        ether a8:5e:45:a1:4c:0b  txqueuelen 1000  (以太网)
        RX packets 472381  bytes 492252080 (492.2 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 318516  bytes 43998888 (43.9 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (本地环回)
        RX packets 643225  bytes 230702679 (230.7 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 643225  bytes 230702679 (230.7 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

virbr0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 192.168.122.1  netmask 255.255.255.0  broadcast 192.168.122.255
        ether 52:54:00:87:8d:e4  txqueuelen 1000  (以太网)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

virbr0-nic: flags=4098<BROADCAST,MULTICAST>  mtu 1500
        ether 52:54:00:87:8d:e4  txqueuelen 1000  (以太网)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wlp4s0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        ether 7c:b2:7d:5e:1e:10  txqueuelen 1000  (以太网)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0



scrapy@node1:~$ docker run --rm -it busybox
/ # ping 172.17.0.1
PING 172.17.0.1 (172.17.0.1): 56 data bytes
^C
--- 172.17.0.1 ping statistics ---
3 packets transmitted, 0 packets received, 100% packet loss
/ # exit

`

**Comment by pyrho** [comment]
@expz 's workaround from March 2018 worked for me...

**Comment by tcassaert** [comment]
What worked for me is loading the `br_netfilter` kernel module:
```
lsmod | grep br_netfilter
modprobe br_netfilter
```
This was on a CentOS 7 node, where I had to upgrade the kernel before I could load the kernel module.

Another possible solution is to run the `docker0` interface in promiscuous mode:
```
# Check if it's in promisc mode (Check for P in the Flg column of the output)
netstat -i
Kernel Interface table
Iface      MTU    RX-OK RX-ERR RX-DRP RX-OVR    TX-OK TX-ERR TX-DRP TX-OVR Flg
docker0   1500  3058281      0      0 0       5478012      0      0      0 BMU
eth0      1400 19935201      0      0 0      19519868      0      0      0 BMRU
lo       65536  2372033      0      0 0       2372033      0      0      0 LRU

ip link set dev docker0 promisc on
```

**Comment by jniggemann** [comment]
Tried most (if not all) of the solutions above with no luck. This suddenly happens on a docker host running Debian Buster (Debian 10.9, Docker 20.10.6, swarm disabled). Using host-networking works, but I found no other solution to be able to ping anything else than the host<->container. Container->Internet and container<->container do not work (default bridge network and specially-created test overlay network). So Connectivity on the host itself works fine, but not via a container with a bridged or overlay network.

**Comment by pndiku** [comment]
> What worked for me is loading the `br_netfilter` kernel module:
> 
> ```
> lsmod | grep br_netfilter
> modprobe br_netfilter
> ```
> 

Backing up the above. I fought with this for a day until I compared with a server where networking worked.

If you're on Centos7 make sure you upgrade your kernel. The default kernel doesn't include the br_netfilter module.

**Comment by mvandeplassche** [comment]
I used to stop nftables before starting docker :

`systemctl stop nftables`
`systemctl start docker`

Disabling nftables and rebooting fixed the issue :

`systemctl disable nftables`
`reboot`
`systemctl start docker`

**Comment by asfour75** [comment]
> I used to stop nftables before starting docker :
> 
> `systemctl stop nftables` `systemctl start docker`
> 
> It stopped working, containers couldn't reach internet.
> 
> Disabling nftables and rebooting fixed the issue :
> 
> `systemctl disable nftables` `reboot` `systemctl start docker`
> 
> Will try downgrading nftables.

This is insane ... I literally wasted 3 days trying to figure this out .. I tried everything .. every single suggestion, reinstalled everything like 10 times .. till I found this page by accident (I was in the WC) .. How come such thing was never mentioned in dokcers' forums the last 7 years ?

Thank you @mvandeplassche you saved my a** ...

**Comment by jcaesar** [comment]
For me, this problem occurred because I was using nftables and had installed my own `inet filter` table for both ipv4 and ipv6. This table default-dropped traffic from docker. Now you'd think that docker installs rules to get around this default drop through `iptables-nft`, but that only affects the `ip filter` table. Both tables need to accept the traffic.

**Comment by lmachacek** [comment]
We also had this problem and in our case the problem was caused by an old kernel in CentOS 7 when running docker in docker. With kernel 3.10.x outbound network access did not work but after upgrade to kernel 5.4.x it started to work.

**Comment by LordMike** [comment]
I'm battling with this with HASS (Home Assistant) after upgrading from Debian 10 to Debian 11. Even comparing a working Debian 10 (which I'm now refusing to upgrade), I can't spot anything in iptables that's different - but I'm fairly certain the issue is there. Reading this thread has given a lot of tips.

I'm currently stuck at my packets being lost in one `bridge` network, but not another. HASS creates a network `hassio` which is a `bridged network`, and leaves the default network `bridge`. In my current situation, I can `ping` the world (and my local network) from `bridge`, but not `hassio`. I'm down to these `iptables` rules in the `FORWARDING` chain - I test using `ping -i0.1 192.168.0.1` in a container on each network, which will rapidly increment the packet counters so I can see where I'm at.

```
# FORWARD chain with two networks "docker0" (bridge) and "hassio"
Chain FORWARD (policy DROP 4114 packets, 451135 bytes)
    pkts      bytes target     prot opt in     out     source               destination
    9100  1261866 DOCKER-USER  all  --  any    any     anywhere             anywhere
    9100  1261866 DOCKER-ISOLATION-STAGE-1  all  --  any    any     anywhere             anywhere
     382   429067 ACCEPT     all  --  any    docker0  anywhere             anywhere             ctstate RELATED,ESTABLISHED
       0        0 DOCKER     all  --  any    docker0  anywhere             anywhere
     415    42756 ACCEPT     all  --  docker0 !docker0  anywhere             anywhere
       0        0 ACCEPT     all  --  docker0 docker0  anywhere             anywhere
    2924   244984 ACCEPT     all  --  any    hassio  anywhere             anywhere             ctstate RELATED,ESTABLISHED
    1265    93924 DOCKER     all  --  any    hassio  anywhere             anywhere
    1245    92692 ACCEPT     all  --  hassio !hassio  anywhere             anywhere
      20     1232 ACCEPT     all  --  hassio hassio  anywhere             anywhere
```

I've noticed that when I do the following:

```
# docker run --rm -it --net bridge busybox ping -i0.1 192.168.0.1
Then the two counters for docker0 will increment rapidly (marked with '---')

 ---   382   429067 ACCEPT     all  --  any    docker0  anywhere             anywhere             ctstate RELATED,ESTABLISHED
       0        0 DOCKER     all  --  any    docker0  anywhere             anywhere
 ---   415    42756 ACCEPT     all  --  docker0 !docker0  anywhere             anywhere
       0        0 ACCEPT     all  --  docker0 docker0  anywhere             anywhere

But conversely, if I do the following:
# docker run --rm -it --net hassio busybox ping -i0.1 192.168.0.1

Only **ONE** of the below rules increment rapidly (marked with '---').
---  2924   244984 ACCEPT     all  --  any    hassio  anywhere             anywhere             ctstate RELATED,ESTABLISHED
    1265    93924 DOCKER     all  --  any    hassio  anywhere             anywhere
    1245    92692 ACCEPT     all  --  hassio !hassio  anywhere             anywhere
      20     1232 ACCEPT     all  --  hassio hassio  anywhere             anywhere
```

If I `tcpdump` on `hassio`, I can see ICMP packets going out, and if I do it on `eth0` (host), I can see ICMP packets going both out and back in. So iptables definitely seems to be doing something.

I've tried tracking down the `DOCKER` chain, even adding `ACCEPT` and `RETURN` jumps, but to no avail - the packets don't seem to disappear in that - so I have no idea why that `ACCEPT ... hassio !hassio` rule wouldn't get hit.

**Comment by stouch** [comment]
`reboot` helped.

**Comment by niklas** [comment]
For me, this was caused by using an alternative ethernet device (a USB-C multiplicator) aside from the build-in ethernet port.

**Comment by LordMike** [comment]
Interesting you should write. I encountered this exact thing again, a few days ago. I resolved it this time though, as I removed `nftables`. I identified, through some post on the internet, that `nft` was available in Debian 11, but not my Debian 10. 

Removing the `nftables` package from Debian 11 seemed to resolve the issue.. Docker seems to insist on using `iptables`.

**Comment by thoni56** [comment]
I've had this of and on (Docker for Windows) used from WSL. 

> In my case, the reboot of the OS probably solved the problem because it killed the VPN connection and restarted the docker service. There is some conflict between my VPN connection and the docker networking setup. You could also try the four methods I listed.

I just notixed that I had a VPN running. Killed it and access to internet from the Docker container started working again.

**Comment by misq007** [comment]
> We also had this problem and in our case the problem was caused by an old kernel in CentOS 7 when running docker in docker. With kernel 3.10.x outbound network access did not work but after upgrade to kernel 5.4.x it started to work.

I can confirm that after upgrading kernel to 6.x from 3.10.x the issue is resolved.

**Comment by stouch** [comment]
In my case, this thread helped: https://github.com/moby/moby/issues/40217

(and finally, because of some unclear issue with my provider OVH, I had to move from Debian12 to Debian10 and it fixed everything)

**Comment by new-php** [comment]
---

I recently faced an issue where my **Docker containers lost internet connectivity**, though inbound connections worked fine. It was puzzling because everything was functioning perfectly before, and there hadn’t been any significant changes to my setup.

### **Symptoms of the Problem**
- Ping from the host to external IPs (like 8.8.8.8) worked fine.
- Inside the containers, pings to the same IPs resulted in 100% packet loss.
- Using the `--net=host` flag temporarily resolved the issue, but it wasn’t a practical solution.

### **Steps I Took to Fix It**
Here’s how I resolved the issue:

1. **Check Kernel Parameters**  
   After some digging, I discovered that the `bridge-nf-call-iptables` settings in my kernel were disabled. Re-enabling them was the first step:
   ```bash
   sudo sysctl net.bridge.bridge-nf-call-iptables=1
   sudo sysctl net.bridge.bridge-nf-call-ip6tables=1
   sudo sysctl -p
   ```
   This was a key fix. Without these parameters enabled, traffic from the Docker bridge (`docker0`) wasn't being routed correctly.

2. **Reset Docker Network**  
   I also reset the Docker network stack to ensure there were no corrupt configurations:
   ```bash
   sudo ip link delete docker0
   sudo systemctl restart docker
   ```

3. **Firewall Rules**  
   It turned out my firewall was blocking traffic from containers. Adding these rules resolved the issue:
   ```bash
   sudo iptables -A FORWARD -i docker0 -o eth0 -j ACCEPT
   sudo iptables -A FORWARD -i eth0 -o docker0 -j ACCEPT
   ```

4. **Consult Resources**  
   This fix wouldn’t have been possible without some excellent resources I found. The  Vultr guide on [Setting Up Docker Containers on Ubuntu](https://docs.vultr.com/how-to-install-docker-on-ubuntu-24-04) gave me insights into kernel parameter issues. Their [firewall management documentation](https://docs.vultr.com/use-cloudflare-and-vultr-firewall-to-protect-a-vultr-cloud-server) also helped me properly configure rules.

### **Outcome**
After applying these changes, my Docker containers regained full internet connectivity, and everything has been stable since. This was a challenging issue to debug, but I learned a lot in the process. If you’re facing a similar problem, I’d highly recommend checking out those guides—they were invaluable in troubleshooting my setup.

---
